### DISCUSSION TITLE
Logs component

### DISCUSSION THREAD

**Discussion by Alexandring** [ArthurSonzogni/FTXUI#400] [Q&A] [answered]
Hello. I'm tempted to use FTXUI for a small program. My program receives messages over the network. I want to display them in a scrollable window. I can't figure out how to do this with FTXUI. Do you have any thoughts on this?

**Comment by ArthurSonzogni** [comment, marked as answer]
There are many ways.
Let's say you are going to store the data into:
```cpp
std::vector<std::string> log;
```

You can display them inside the `Menu` component:
```cpp
int log_index = 0;
auto log_menu = Menu(&log, &log_index)
```

To make the menu displayed inside a scrollable area, you can use the element decorator: `frame`. You can also add a scroll indicator using: `vscroll_indicator` decorator:

```cpp
auto log_menu_decorated = log_menu | vscroll_indicator | frame;
```

<details>
  <summary>Alternatively</summary>
  
```cpp
auto log_menu_decorated = Renderer(log_menu, [log_menu] {
  return log_menu->Render() | vscroll_indicator | frame;
});
```
</details>

That's it. 
If you don't want to use `Menu`, you should be able to build any kind of component rendering the log the way you want, and then wrap them inside `Scroller` from:
https://github.com/ArthurSonzogni/git-tui/blob/master/src/scroller.cpp

One final thing to note. When calling `ScreenInteractive::Loop`, the program is running the FTXUI main loop. It waits for events, and then draw a new frame. It doesn't know about network events. You need to post new events to get new frame to be drawn.

For instance you can post from a thread:
```cpp
screen_interactive.PostEvent(Event::Custom);
```
this will force a new frame to be drawn.

If the thread waiting for network and the one running FTXUI aren't the same, you can also use ScreenInteractive::Post to post task from one thread to the other:
```cpp
  void AddLogFromNetworkThread(std::string line) {
    // Currently on the "network" thread.
    screen_.PostTask([line]() {
      // Currently on the "UI" thread.
      log.push_back(line);
      screen_.PostEvent(Event::Custom); // Cause a new frame to be drawn.
    });
  }
```

**Reply by Chasikanaft** [reply]
Can you show how correctly launch and send messages to UI log in another thread?

**Reply by ArthurSonzogni** [reply]
I think I just did.

"PostTask" and "PostEvent" are thread safe. They are what "send" message from the thread toward the UI thread.

Internally, there are using the Sender/Receiver class:
https://github.com/ArthurSonzogni/FTXUI/blob/master/include/ftxui/component/receiver.hpp

**Reply by Chasikanaft** [reply]
I have no functions "PostTask" and "PostEvent" in "ScreenInteractive" class :) . Cause i still have no update after 2.0. I can do something with vcpkg or i must build libs manualy?

**Reply by ArthurSonzogni** [reply]
CMake FetchContent is the best way IMO.
Alternatively, I am going to update the vcpkg version later today. You can wait for this.

**Reply by ArthurSonzogni** [reply]
VCPKG port update:
https://github.com/microsoft/vcpkg/pull/25063

**Reply by Chasikanaft** [reply]
Thanks!
